File: mcp-code-intelligence-python/src/mcp_code_intel/analyzers/security/cfg/control_flow_graph.py
"""
KSA-164: Control Flow Graph — Container for basic blocks and edges.
"""

from __future__ import annotations
from typing import Optional
from .basic_block import BasicBlock
from .cfg_edge import CFGEdge


class ControlFlowGraph:
    def __init__(self, entry: BasicBlock) -> None:
        self.entry = entry
        self.exits: list[BasicBlock] = []
        self.blocks: list[BasicBlock] = []
        self.edges: list[CFGEdge] = []
        self._adjacency: dict[int, list[CFGEdge]] = {}
        self._reverse_adj: dict[int, list[CFGEdge]] = {}
        self.add_block(entry)

    def add_block(self, block: BasicBlock) -> None:
        self.blocks.append(block)
        self._adjacency[block.id] = []
        self._reverse_adj[block.id] = []
        if block.type == "exit":
            self.exits.append(block)

    def add_edge(self, from_block: BasicBlock, to_block: BasicBlock, edge_type: str, label: Optional[str] = None) -> CFGEdge:
        edge = CFGEdge(from_block, to_block, edge_type, label)
        self.edges.append(edge)
        self._adjacency[from_block.id].append(edge)
        self._reverse_adj[to_block.id].append(edge)
        return edge

    def get_successors(self, block: BasicBlock) -> list[BasicBlock]:
        return [e.to_block for e in self._adjacency.get(block.id, [])]

    def get_predecessors(self, block: BasicBlock) -> list[BasicBlock]:
        return [e.from_block for e in self._reverse_adj.get(block.id, [])]

    def get_out_edges(self, block: BasicBlock) -> list[CFGEdge]:
        return self._adjacency.get(block.id, [])

    def get_in_edges(self, block: BasicBlock) -> list[CFGEdge]:
        return self._reverse_adj.get(block.id, [])

    def topological_order(self) -> list[BasicBlock]:
        visited: set[int] = set()
        result: list[BasicBlock] = []

        def dfs(block: BasicBlock) -> None:
            if block.id in visited:
                return
            visited.add(block.id)
            for succ in self.get_successors(block):
                dfs(succ)
            result.insert(0, block)

        dfs(self.entry)
        return result

    def reverse_post_order(self) -> list[BasicBlock]:
        visited: set[int] = set()
        post_order: list[BasicBlock] = []

        def dfs(block: BasicBlock) -> None:
            if block.id in visited:
                return
            visited.add(block.id)
            for succ in self.get_successors(block):
                dfs(succ)
            post_order.append(block)

        dfs(self.entry)
        return list(reversed(post_order))

    def get_block(self, block_id: int) -> Optional[BasicBlock]:
        for b in self.blocks:
            if b.id == block_id:
                return b
        return None

    def __str__(self) -> str:
        lines = [f"CFG: {len(self.blocks)} blocks, {len(self.edges)} edges"]
        for edge in self.edges:
            lines.append(f"  {edge}")
        return "\n".join(lines)

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/security/cfg/ControlFlowGraph.kt
/**
 * KSA-164: Control Flow Graph — Container for basic blocks and edges.
 */
package com.codeintel.analyzers.security.cfg

import com.codeintel.analyzers.security.EdgeType

class ControlFlowGraph(val entry: BasicBlock) {
    val exits: MutableList<BasicBlock> = mutableListOf()
    val blocks: MutableList<BasicBlock> = mutableListOf()
    val edges: MutableList<CFGEdge> = mutableListOf()
    private val adjacency: MutableMap<Int, MutableList<CFGEdge>> = mutableMapOf()
    private val reverseAdj: MutableMap<Int, MutableList<CFGEdge>> = mutableMapOf()

    init {
        addBlock(entry)
    }

    fun addBlock(block: BasicBlock) {
        blocks.add(block)
        adjacency[block.id] = mutableListOf()
        reverseAdj[block.id] = mutableListOf()
        if (block.type == com.codeintel.analyzers.security.BlockType.EXIT) exits.add(block)
    }

    fun addEdge(from: BasicBlock, to: BasicBlock, type: EdgeType, label: String? = null): CFGEdge {
        val edge = CFGEdge(from, to, type, label)
        edges.add(edge)
        adjacency[from.id]!!.add(edge)
        reverseAdj[to.id]!!.add(edge)
        return edge
    }

    fun getSuccessors(block: BasicBlock): List<BasicBlock> =
        adjacency[block.id]?.map { it.to } ?: emptyList()

    fun getPredecessors(block: BasicBlock): List<BasicBlock> =
        reverseAdj[block.id]?.map { it.from } ?: emptyList()

    fun getOutEdges(block: BasicBlock): List<CFGEdge> =
        adjacency[block.id] ?: emptyList()

    fun getInEdges(block: BasicBlock): List<CFGEdge> =
        reverseAdj[block.id] ?: emptyList()

    fun topologicalOrder(): List<BasicBlock> {
        val visited = mutableSetOf<Int>()
        val result = mutableListOf<BasicBlock>()

        fun dfs(block: BasicBlock) {
            if (block.id in visited) return
            visited.add(block.id)
            for (succ in getSuccessors(block)) dfs(succ)
            result.add(0, block)
        }

        dfs(entry)
        return result
    }

    fun reversePostOrder(): List<BasicBlock> {
        val visited = mutableSetOf<Int>()
        val postOrder = mutableListOf<BasicBlock>()

        fun dfs(block: BasicBlock) {
            if (block.id in visited) return
            visited.add(block.id)
            for (succ in getSuccessors(block)) dfs(succ)
            postOrder.add(block)
        }

        dfs(entry)
        return postOrder.reversed()
    }

    fun getBlock(id: Int): BasicBlock? = blocks.find { it.id == id }

    override fun toString(): String {
        val lines = mutableListOf("CFG: ${blocks.size} blocks, ${edges.size} edges")
        for (edge in edges) lines.add("  $edge")
        return lines.joinToString("\n")
    }
}
